File: docker/Dockerfile.ts
FROM node:24-alpine AS build-base
COPY hello-websocket-ts /app/hello-websocket-ts
WORKDIR /app/hello-websocket-ts
RUN npm ci

FROM node:24-alpine AS server
WORKDIR /app
COPY --from=build-base /app/hello-websocket-ts/package*.json /app/
COPY --from=build-base /app/hello-websocket-ts/node_modules /app/node_modules
COPY --from=build-base /app/hello-websocket-ts/server /app/server
COPY --from=build-base /app/hello-websocket-ts/client /app/client
COPY --from=build-base /app/hello-websocket-ts/common /app/common
COPY --from=build-base /app/hello-websocket-ts/tsconfig.json /app/
ENTRYPOINT ["node", "--loader", "ts-node/esm", "server/ws_server.ts"]

FROM node:24-alpine AS client
WORKDIR /app
COPY --from=build-base /app/hello-websocket-ts/package*.json /app/
COPY --from=build-base /app/hello-websocket-ts/node_modules /app/node_modules
COPY --from=build-base /app/hello-websocket-ts/server /app/server
COPY --from=build-base /app/hello-websocket-ts/client /app/client
COPY --from=build-base /app/hello-websocket-ts/common /app/common
COPY --from=build-base /app/hello-websocket-ts/tsconfig.json /app/
ENTRYPOINT ["node", "--loader", "ts-node/esm", "client/ws_client.ts"]

File: docker/Dockerfile.java
# hello-websocket-java/pom.xml sets maven.compiler.source/target=17, the
# bytecode target this module compiles to. Building and running on a newer
# JDK (25 here) is standard Java forward compatibility (javac --release 17
# semantics) and is not a version mismatch to "fix" by pinning this image
# to JDK 17.
FROM maven:3-eclipse-temurin-25 AS build-base
COPY docker/settings.xml /root/.m2/settings.xml
WORKDIR /app
COPY hello-websocket-java/pom.xml /app/pom.xml
COPY hello-websocket-java/common /app/common
COPY hello-websocket-java/server /app/server
COPY hello-websocket-java/client /app/client
RUN mvn --batch-mode clean package
RUN cp target/hello-websocket-java-1.0.0.jar /app/hello-ws.jar

FROM eclipse-temurin:25-jre-alpine AS server
RUN addgroup -S app && adduser -S -G app app
WORKDIR /app
COPY --from=build-base --chown=app:app /app/hello-ws.jar /app/
USER app
ENTRYPOINT ["java", "-jar", "hello-ws.jar"]

FROM eclipse-temurin:25-jre-alpine AS client
RUN addgroup -S app && adduser -S -G app app
WORKDIR /app
COPY --from=build-base --chown=app:app /app/hello-ws.jar /app/
USER app
ENTRYPOINT ["java", "-cp", "hello-ws.jar", "io.github.hellowebsocket.WsClient"]
